docs/settings.md
Settings file
=============

The settings file for Thirty is an ini file, such as:

```ini
[server]
port = 25565
name = Sean test server
motd = Things will break and go boom.
public = true
offline = false
max_players = 16
whitelist = false

[map]
name = world4
width = 64
depth = 64
height = 64
generator = classic
seed = 2015

[colours]
x = 0099ff33
y = aabbccdd
```

## Location

By default, settings are loaded from `settings.ini` in the working directory of the server.
You can use the `-c` command-line parameter to specify a file to use:

```
thirty -c /path/to/custom/settings.ini
```

## Fields
### `server`

| Key                      | Default                                                 | Description                                                                                                                                                                             |
|--------------------------|---------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **`port`**               | 25565                                                   | TCP port the server will listen on.                                                                                                                                                     |
| **`name`**               | `Unnamed server`                                        | Name of the server, shown on the server list and to connecting clients.                                                                                                                 |
| **`motd`**               | `The server owner needs to set a MotD in settings.ini.` | Message of the day, shown to connecting clients.                                                                                                                                        |
| **`max_players`**        | 8                                                       | Maximum number of players that can be connected at one time.                                                                                                                            |
| **`public`**             | true                                                    | Whether this server will be displayed on the server list.                                                                                                                               |
| **`offline`**            | true                                                    | If enabled, this disables name verification, and disables heartbeat entirely.                                                                                                           |
| **`whitelist`**          | false                                                   | Whether the whitelist is enabled, preventing users not listed in `whitelist.txt` from connecting.                                                                                       |
| **`enable_old_clients`** | false                                                   | Allows clients using protocol versions than `7` (classic 0.30) to connect. Experimental; those clients may behave unreliably.                                                           |
| **`web_proxies`**        | `34.223.5.250`                                          | IP addresses allowed to act as a WebSocket proxy. These IPs can pretend to be from any other IP. The default value is [ClassiCube's proxy](https://www.classicube.net/api/docs/server). |
| **`heartbeat_url`**      | `https://www.classicube.net/server/heartbeat/`          | URL of the heartbeat API used to broadcast the server to a server list.                                                                                                                 |

### `map`

| Key                  | Default   | Description                                                                                                           |
|----------------------|-----------|-----------------------------------------------------------------------------------------------------------------------|
| **`name`**           | `world`   | Name of the world.                                                                                                    |
| **`width`**          | 64        | Width of the world (X axis size)                                                                                      |
| **`depth`**          | 64        | Depth of the world (Y axis size)                                                                                      |
| **`height`**         | 64        | Height of the world (Z axis size)                                                                                     |
| **`generator`**      | `classic` | The level generator to use, see [Level generators](map_generators.md).                                                |
| **`seed`**           | 0         | Seed for the random number generator used by the level generator.                                                     |
| **`image_path`**     |           | If this and `image_interval` are specified, an image of the map will be generated at this location.                   |
| **`image_interval`** | 0         | If this and `image_path` are specified, an image of the map will be generated periodically. This value is in seconds. |

### `colours`

Defines custom text colours for use with the [`TextColors`](https://wiki.vg/Classic_Protocol_Extension#TextColors) CPE extension.
The alternative spelling `colors` is also allowed for this section.

The key will be the colour code, the value is an 8-digit hexadecimal number in `RRGGBBAA` format defining the colour.

### `network`

| Key                   | Default | Description                                                                                                                                                                          |
|-----------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **`map_compression`** | `9`     | DEFLATE compression level used when sending the map to clients. This can be between 1-9, where 1 is the fastest and 9 has the best compression. 0 can be used to use no compression. |

### `debug`

Settings under this section are mostly only useful for developing or testing the server.
Avoid using them on actual production servers.

| Key                | Default | Description                                                                                                                                                                         |
|--------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **`fixed_salt`**   |         | Specifies a fixed salt that will be used for client authentication. Usually, a salt is randomly generated at server startup. **Do not use this** except for debugging login issues. |
| **`disable_save`** | `false` | Disables level saving. Mostly useful for writing world generators.                                                                                                                  |


meson_options.txt
option('use_readline', type: 'feature', value: 'auto')


src/config.c
// Thirty, a ClassiCube (Minecraft Classic) server
// Copyright (C) 2024 Sean Baggaley
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <sys/types.h>
#include <errno.h>
#include <stdbool.h>
#include <limits.h>
#include <unistd.h>
#include "config.h"
#include "log.h"

typedef void (*configcallback_t)(const char *section, const char *key, const char *value);

enum parsestate_e {
	parse_section,
	parse_key,
	parse_value,
	parse_comment,
};

static void config_parse(const char *filename, configcallback_t callback);
static long parse_int(const char *ptr, bool *ok, int radix);
static void cfg_callback(const char *section, const char *key, const char *value);

config_t config;

static const char *config_default_paths[] = {
	"/etc/thirty.ini",
	"/usr/local/etc/thirty.ini",
	"thirty.ini",

	NULL
};

static void trim(char *p) {
	size_t len = strlen(p);
	size_t start = 0;
	for (; start < len; start++) {
		if (!isspace(p[start])) {
			break;
		}
	}

	memmove(p, p + start, len - start);
	p[len - start] = 0;
	len = strlen(p);

	for (ssize_t i = (ssize_t)len - 1; i >= 0; i--) {
		if (isspace(p[i])) {
			p[i] = 0;
		}
		else {
			break;
		}
	}
}

void config_parse(const char *filename, configcallback_t callback) {
	log_printf(log_info, "Loading configuration from '%s'.", filename);

	FILE *fp = fopen(filename, "r");
	if (fp == NULL) {
		log_printf(log_error, "Failed to open config file '%s' for reading.", filename);
		return;
	}

	enum parsestate_e state = parse_key;
	char sectionbuf[128];
	size_t sectionp = 0;
	char keybuf[128];
	size_t keyp = 0;
	char valuebuf[1024];
	size_t valuep = 0;

	while (!feof(fp)) {
		char c;
		size_t r = fread(&c, sizeof(char), 1, fp);
		if (r != 1) {
			if (ferror(fp)) {
				log_printf(log_error, "Error while parsing config file '%s': %s", filename, strerror(errno));
			}
			break;
		}

		switch (state) {
			case parse_section: {
				if (c == ']') {
					state = parse_key;
					continue;
				}

				if (sectionp >= sizeof(sectionbuf) - 1) {
					log_printf(log_error, "Section name is too long and will be truncated: '%s'", sectionbuf);
					break;
				}

				sectionbuf[sectionp++] = c;
				sectionbuf[sectionp] = 0;

				break;
			}

			case parse_key: {
				if (keyp == 0) {
					if (c == '[') {
						state = parse_section;
						sectionp = 0;
						continue;
					}
					else if (isspace(c)) {
						continue;
					}
					else if (c == ';' || c == '#') {
						state = parse_comment;
						continue;
					}
				}
				else {
					if (c == '=') {
						state = parse_value;
						valuep = 0;
						continue;
					}
				}

				if (keyp >= sizeof(keybuf) - 1) {
					log_printf(log_error, "Key name is too long and will be truncated: '%s'", keybuf);
					break;
				}

				keybuf[keyp++] = c;
				keybuf[keyp] = 0;

				break;
			}

			case parse_value: {
				if (c == '\n') {
					trim(sectionbuf);
					trim(keybuf);
					trim(valuebuf);
					callback(sectionbuf, keybuf, valuebuf);
					state = parse_key;
					keyp = 0;
					break;
				}
				else if (valuep == 0) {
					if (isspace(c)) {
						continue;
					}
				}

				if (valuep >= sizeof(valuebuf) - 1) {
					log_printf(log_error, "Value is too long and will be truncated: '%s'", valuebuf);
					break;
				}

				valuebuf[valuep++] = c;
				valuebuf[valuep] = 0;

				break;
			}

			case parse_comment: {
				if (c == '\n') {
					state = parse_key;
				}

				break;
			}
		}
	}

	fclose(fp);
}

long parse_int(const char *ptr, bool *ok, int radix) {
	char *end;
	long value = strtol(ptr, &end, radix);
	if (end == ptr || (errno == ERANGE && (value == LONG_MIN || value == LONG_MAX))) {
		*ok = false;
		return 0;
	}
	*ok = true;
	return value;
}

unsigned long parse_uint(const char *ptr, bool *ok, int radix) {
	char *end;
	unsigned long value = strtoul(ptr, &end, radix);
	if (end == ptr || (errno == ERANGE && value == ULONG_MAX)) {
		*ok = false;
		return 0;
	}
	*ok = true;
	return value;
}

void cfg_callback(const char *section, const char *key, const char *value) {
	bool ok;

	if (strcmp(section, "server") == 0) {
		if (strcmp(key, "name") == 0) {
			config.server.name = strdup(value);
		}
		else if (strcmp(key, "motd") == 0) {
			config.server.motd = strdup(value);
		}
		else if (strcmp(key, "port") == 0) {
			long port = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'port' as unsigned integer");
			} else {
				config.server.port = (uint16_t) port;
			}
		}
		else if (strcmp(key, "public") == 0) {
			config.server.public = strcmp(value, "true") == 0;
		}
		else if (strcmp(key, "offline") == 0) {
			config.server.offline = strcmp(value, "true") == 0;
		}
		else if (strcmp(key, "max_players") == 0) {
			long max = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'max_players' as unsigned integer");
			} else {
				config.server.max_players = (unsigned int) max;
			}
		}
		else if (strcmp(key, "whitelist") == 0) {
			config.server.enable_whitelist = strcmp(value, "true") == 0;
		}
		else if (strcmp(key, "web_proxies") == 0) {
			for (size_t i = 0; i < config.server.num_proxies; i++) {
				free(config.server.allowed_web_proxies[i]);
			}
			
			free(config.server.allowed_web_proxies);
			config.server.allowed_web_proxies = NULL;
			config.server.num_proxies = 0;
			
			char buf[64];
			size_t bufp = 0;
			const size_t vallen = strlen(value);
			for (size_t i = 0; i <= vallen; i++) {
				const char c = value[i];
				if (c == ' ' || i == vallen) {
					if (bufp == 0) {
						continue;
					}

					trim(buf);
					size_t idx = config.server.num_proxies++;
					config.server.allowed_web_proxies = realloc(config.server.allowed_web_proxies, sizeof(char *) * config.server.num_proxies);
					config.server.allowed_web_proxies[idx] = strdup(buf);
					bufp = 0;
				}
				else if (bufp <= sizeof(buf) - 1) {
					buf[bufp++] = c;
					buf[bufp] = '\0';
				}
			}
		}
		else if (strcmp(key, "enable_old_clients") == 0) {
			config.server.enable_old_clients = strcmp(value, "true") == 0;
		}
		else if (strcmp(key, "heartbeat_url") == 0) {
			config.server.heartbeat_url = strdup(value);
		}
	}

	else if (strcmp(section, "map") == 0) {
		if (strcmp(key, "name") == 0) {
			config.map.name = strdup(value);
		}
		else if (strcmp(key, "width") == 0) {
			long size = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'width' as unsigned integer");
			} else {
				config.map.width = size;
			}
		}
		else if (strcmp(key, "depth") == 0) {
			long size = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'depth' as unsigned integer");
			} else {
				config.map.depth = size;
			}
		}
		else if (strcmp(key, "height") == 0) {
			long size = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'height' as unsigned integer");
			} else {
				config.map.height = size;
			}
		}
		else if (strcmp(key, "generator") == 0) {
			config.map.generator = strdup(value);
		}
		else if (strcmp(key, "seed") == 0) {
			long seed = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'seed' as unsigned integer");
			} else {
				config.map.seed = seed;
				config.map.random_seed = false;
			}
		}
		else if (strcmp(key, "image_path") == 0) {
			config.map.image_path = strdup(value);
		}
		else if (strcmp(key, "image_interval") == 0) {
			long size = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'image_interval' as unsigned integer");
			} else {
				config.map.image_interval = size;
			}
		}
	}

	else if (strcmp(section, "colours") == 0 || strcmp(section, "colors") == 0) {
		if (strlen(key) != 1) {
			log_printf(log_error, "Colour name must be exactly 1 character.");
			return;
		}
		if (strlen(value) != 8) {
			log_printf(log_error, "Colour value must be an 8-digit RGBA hex number.");
			return;
		}

		uint32_t colour = (uint32_t) parse_uint(value, &ok, 16);
		if (!ok) {
			log_printf(log_error, "Colour value failed to parse, it must be an 8-digit RGBA hex number.");
			return;
		}

		size_t idx = config.num_colours++;
		config.colours = realloc(config.colours, sizeof(*config.colours) * config.num_colours);
		config.colours[idx].code = key[0];
		config.colours[idx].r = (colour >> 24U) & 0xFFU;
		config.colours[idx].g = (colour >> 16U) & 0xFFU;
		config.colours[idx].b = (colour >> 8U) & 0xFFU;
		config.colours[idx].a = colour & 0xFFU;
	}

	else if (strcmp(section, "debug") == 0) {
		if (strcmp(key, "fixed_salt") == 0) {
			strncpy(config.debug.fixed_salt, value, sizeof(config.debug.fixed_salt) - 1);
		}
		else if (strcmp(key, "disable_save") == 0) {
			config.debug.disable_save = strcmp(value, "true") == 0;
		}
		else if (strcmp(key, "dump_connection_packets") == 0) {
			config.debug.dump_connection_packets = strcmp(value, "true") == 0;
		}
	}

	else if (strcmp(section, "network") == 0) {
		if (strcmp(key, "map_compression") == 0) {
			long level = parse_int(value, &ok, 10);
			if (!ok) {
				log_printf(log_error, "Failed to parse 'map_compression' as integer");
			} else {
				if (level < 0) level = 0;
				if (level > 9) level = 9;
				config.network.map_compression = (int)level;
			}
		}
	}
}

void config_init(const char *config_path) {
	memset(&config, 0, sizeof(config));

	config.map.random_seed = true;

	config.server.allowed_web_proxies = malloc(sizeof(char *));
	config.server.allowed_web_proxies[0] = strdup("34.223.5.250");
	config.server.num_proxies = 1;

	config.network.map_compression = 9;

	if (config_path != NULL) {
		config_parse(config_path, cfg_callback);
	}
	else {
		for (size_t i = 0; config_default_paths[i] != NULL; i++) {
			if (access(config_default_paths[i], R_OK) != 0) {
				continue;
			}

			config_parse(config_default_paths[i], cfg_callback);
			break;
		}
	}

	if (config.server.name == NULL) {
		config.server.name = strdup("Unnamed server");
	}

	if (config.server.motd == NULL) {
		config.server.motd = strdup("The server owner needs to set a MotD in settings.ini.");
	}

	if (config.server.port == 0) {
		config.server.port = 25565;
	}

	if (config.server.max_players == 0) {
		config.server.max_players = 8;
	}

	if (config.server.heartbeat_url == NULL) {
		config.server.heartbeat_url = strdup("https://www.classicube.net/server/heartbeat/");
	}

	if (config.map.name == NULL) {
		config.map.name = strdup("world");
	}

	if (config.map.width == 0) {
		config.map.width = 64;
	}

	if (config.map.depth == 0) {
		config.map.depth = 64;
	}

	if (config.map.height == 0) {
		config.map.height = 64;
	}

	if (config.map.generator == NULL) {
		config.map.generator = strdup("classic");
	}

	if (config.map.image_path == NULL) {
		config.map.image_path = strdup("");
	}
}

void config_destroy(void) {
	for (size_t i = 0; i < config.server.num_proxies; i++) {
		free(config.server.allowed_web_proxies[i]);
	}

	free(config.server.allowed_web_proxies);
	config.server.allowed_web_proxies = NULL;
	config.server.num_proxies = 0;
	
	free(config.server.name);
	free(config.server.motd);
	free(config.server.heartbeat_url);
	free(config.map.image_path);
	free(config.map.name);
	free(config.map.generator);

	memset(&config, 0, sizeof(config));
}

const textcolour_t *config_find_colour(const char code) {
	for (size_t i = 0; i < config.num_colours; i++) {
		if (config.colours[i].code == code) {
			return &config.colours[i];
		}
	}

	return NULL;
}
